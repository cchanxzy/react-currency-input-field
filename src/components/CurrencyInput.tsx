import React, { FC, useState } from 'react';

import { addCommas, checkIsValidNumber, removeCommas } from './utilities';

interface IProps {
  id: string;
  className?: string;
  onChange: (value: number | null) => void;
  placeholder?: string;
  prefix?: string;
}

export const CurrencyInput: FC<IProps> = ({ id, className, onChange, placeholder, prefix }) => {
  const [value, setValue] = useState('');

  const processChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const {
      target: { value: eventVal },
    } = event;

    let stringValue = eventVal;

    if (prefix) {
      stringValue = stringValue.replace(prefix, '');
    }

    if (stringValue === '' || stringValue === null) {
      onChange(null);
      return setValue('');
    }

    const intValue = parseInt(removeCommas(stringValue), 10);

    if (checkIsValidNumber(intValue)) {
      let newValue = addCommas(intValue);

      if (prefix) {
        newValue = `${prefix}${newValue}`;
      }

      setValue(newValue);
    }

    onChange(intValue);
  };

  return (
    <input
      type="string"
      id={id}
      className={className}
      onChange={processChange}
      placeholder={placeholder}
      value={value}
    />
  );
};

export default CurrencyInput;
