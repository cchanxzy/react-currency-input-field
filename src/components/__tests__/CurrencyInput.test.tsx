import { shallow } from 'enzyme';
import React from 'react';
import CurrencyInput from '../CurrencyInput';

describe('<CurrencyInput /> component', () => {
  it('Renders without error', () => {
    const id = 'validationCustom01';
    const placeholder = '£1,000';
    const className = 'form-control';

    const view = shallow(
      <CurrencyInput id={id} placeholder={placeholder} className={className} onChange={jest.fn()} />
    );
    const input = view.find('#validationCustom01');
    expect(input).toMatchSnapshot();

    expect(input.prop('id')).toBe(id);
    expect(input.prop('placeholder')).toBe(placeholder);
    expect(input.prop('className')).toBe(className);
  });

  it('should allow input change value', () => {
    const onChange = jest.fn();
    const view = shallow(
      <CurrencyInput
        id="validationCustom01"
        placeholder="£1,000"
        className="form-control"
        prefix="£"
        onChange={onChange}
      />
    );
    const input = view.find('#validationCustom01');
    input.simulate('change', { target: { value: '100' } });

    expect(onChange).toBeCalledWith(100);
  });

  it('should allow empty value change', () => {
    const onChange = jest.fn();
    const view = shallow(
      <CurrencyInput
        id="validationCustom01"
        placeholder="£1,000"
        className="form-control"
        prefix="£"
        onChange={onChange}
      />
    );
    const input = view.find('#validationCustom01');
    input.simulate('change', { target: { value: '' } });

    expect(onChange).toBeCalledWith(null);
  });

  it('should not allow string value change', () => {
    const onChange = jest.fn();
    const view = shallow(
      <CurrencyInput
        id="validationCustom01"
        placeholder="£1,000"
        className="form-control"
        prefix="£"
        onChange={onChange}
      />
    );
    const input = view.find('#validationCustom01');
    input.simulate('change', { target: { value: 'aakoa' } });

    expect(onChange).toBeCalledWith(NaN);
  });
});
