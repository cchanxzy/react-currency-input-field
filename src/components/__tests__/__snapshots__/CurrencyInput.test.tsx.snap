// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CurrencyInput /> component Renders without error 1`] = `ShallowWrapper {}`;
